### PR TITLE
Improves TaskContext#isActive checks on queries

### PR DESCRIPTION
**Pull request by mkeckmkeck**
We want to automatically stop:
- iterations without limits (means e.g. iterate, streamBlockwise, scrolling)
- and deletions
in case the task context is no longer active
But "short-running" request should still run through.

### CODE REVIEW

**Review by qw3ry — APPROVED** [review]

**Review by idlira — APPROVED** [review]

**Review by fhaScireum — APPROVED** [review]

**Review by jakobvogel — APPROVED** [review]

File: src/main/java/sirius/db/jdbc/BaseSQLQuery.java [hunk 1]
```diff
@@ -74,7 +93,7 @@ public List<Row> queryList(Limit limit) throws SQLException {
      *                that there is no limit.
      * @throws SQLException in case of a database error
      */
-    public abstract void iterate(Predicate<Row> handler, @Nullable Limit limit) throws SQLException;
+    public abstract void doIterate(Predicate<Row> handler, @Nullable Limit limit) throws SQLException;
```
Comment (andyHa) [inline src/main/java/sirius/db/jdbc/BaseSQLQuery.java:96]: protected?

File: src/main/java/sirius/db/es/ElasticQuery.java [hunk 2]
```diff
@@ -1048,7 +1048,7 @@ public boolean exists() {
 
     @SuppressWarnings("unchecked")
     @Override
-    public void iterate(Predicate<E> handler) {
+    public void doIterate(Predicate<E> handler) {
```
Comment (andyHa) [inline src/main/java/sirius/db/es/ElasticQuery.java:1051]: protected?

File: src/main/java/sirius/db/jdbc/BaseSQLQuery.java [hunk 3]
```diff
@@ -58,14 +58,33 @@ public List<Row> queryList() throws SQLException {
...
+     * Consider using the method instead of {@link #queryList()} if a large result set is expected as this method. As
+     * this method only processes one row at a time, this might be much more memory efficient.
+     *
+     * @param handler the row handler invoked for each row
+     * @param limit   the limit which controls which and how many rows are output. Can be <tt>null</tt> to indicate
+     *                that there is no limit.
+     * @throws SQLException in case of a database error
+     */
+    public void iterate(Predicate<Row> handler, @Nullable Limit limit) throws SQLException {
+        doIterate(handler, limit);
```
Comment (andyHa) [inline src/main/java/sirius/db/jdbc/BaseSQLQuery.java:79]: shouldn't the TaskContext be checked here?

File: src/main/java/sirius/db/jdbc/SQLQuery.java [hunk 4]
```diff
@@ -108,22 +107,21 @@ public SQLQuery markAsLongRunning() {
     }
 
     @Override
-    public void iterate(Predicate<Row> handler, @Nullable Limit limit) throws SQLException {
-        Watch w = Watch.start();
+    public void doIterate(Predicate<Row> handler, @Nullable Limit limit) throws SQLException {
```
Comment (andyHa) [inline src/main/java/sirius/db/jdbc/SQLQuery.java:110]: protected

File: src/main/java/sirius/db/jdbc/SmartQuery.java [hunk 5]
```diff
@@ -456,7 +457,7 @@ public SmartQuery<E> copy() {
     }
 
     @Override
-    public void iterate(Predicate<E> handler) {
+    public void doIterate(Predicate<E> handler) {
```
Comment (andyHa) [inline src/main/java/sirius/db/jdbc/SmartQuery.java:460]: proteted

File: src/main/java/sirius/db/jdbc/TransformedQuery.java [hunk 6]
```diff
@@ -37,7 +37,7 @@ protected TransformedQuery(EntityDescriptor descriptor, String alias, SQLQuery q
     }
 
     @Override
-    public void iterate(Predicate<E> handler) {
+    public void doIterate(Predicate<E> handler) {
```
Comment (andyHa) [inline src/main/java/sirius/db/jdbc/TransformedQuery.java:40]: protected

File: src/main/java/sirius/db/jdbc/batch/BatchSQLQuery.java [hunk 7]
```diff
@@ -32,13 +31,12 @@ protected BatchSQLQuery(CustomQuery customQuery) {
     }
 
     @Override
-    public void iterate(Predicate<Row> handler, @Nullable Limit limit) throws SQLException {
-        Watch w = Watch.start();
+    public void doIterate(Predicate<Row> handler, @Nullable Limit limit) throws SQLException {
```
Comment (andyHa) [inline src/main/java/sirius/db/jdbc/batch/BatchSQLQuery.java:34]: protected

File: src/main/java/sirius/db/jdbc/batch/external/ExternalBatchQuery.java [hunk 8]
```diff
@@ -140,10 +139,9 @@ public void addBatch() throws SQLException {
     }
 
     @Override
-    public void iterate(Predicate<Row> handler, @Nullable Limit limit) throws SQLException {
-        try (ResultSet rs = statement.executeQuery()) {
-            TaskContext tc = TaskContext.get();
-            processResultSet(handler, limit, rs, tc);
+    public void doIterate(Predicate<Row> handler, @Nullable Limit limit) throws SQLException {
```
Comment (andyHa) [inline src/main/java/sirius/db/jdbc/batch/external/ExternalBatchQuery.java:142]: protected

File: src/main/java/sirius/db/mixing/query/BaseQuery.java [hunk 9]
```diff
@@ -224,8 +226,25 @@ public Stream<E> streamList() {
...
+     * Note that this method is intended for medium-sized results as not all items in the result need to be
+     * kept in memory when iterating through them. Note however, that for very large result sets, a method
+     * like {@link #streamBlockwise()} might be more appropriate, as it ensures that underlying resources
+     * like <tt>cursors</tt> or <tt>database connections</tt> cannot run into a timeout.
+     *
+     * @param handler the handler to be invoked for each item in the result. Should return <tt>true</tt>
+     *                to continue processing or <tt>false</tt> to abort processing of the result set.
+     */
+    public void iterate(Predicate<E> handler) {
+        doIterate(handler.and(ignored -> TaskContext.get().isActive()));
```
Comment (andyHa) [inline src/main/java/sirius/db/mixing/query/BaseQuery.java:242]: get TaskContext outside of the "loop" so save some us?

File: src/main/java/sirius/db/jdbc/BaseSQLQuery.java [hunk 10]
```diff
@@ -133,8 +150,10 @@ public Optional<Row> first() throws SQLException {
     @Nullable
     public Row queryFirst() throws SQLException {
         ValueHolder<Row> result = ValueHolder.of(null);
-        iterateAll(result, Limit.singleItem());
-
+        doIterate(row -> {
+            result.accept(row);
+            return true;
```
Comment (andyHa) [inline src/main/java/sirius/db/jdbc/BaseSQLQuery.java:155]: we could actually also return false here as there shouldn't be anything left in the result set anyway...

**Review by andyHa — CHANGES_REQUESTED** [review]

File: src/main/java/sirius/db/jdbc/BaseSQLQuery.java (hunk 3 above)
Comment (mkeckmkeck) [reply src/main/java/sirius/db/jdbc/BaseSQLQuery.java]: 🙈

**Review by fhaScireum — APPROVED** [review]

**Review by idlira — APPROVED** [review]

**Review by mko-sci — APPROVED** [review]

**Review by andyHa — APPROVED** [review]

**Review by sabieber — APPROVED** [review]

**Pull request merged**